File: appMain/src/main/java/com/cartlc/tracker/data/DataProjectAddressCombo.java
package com.cartlc.tracker.data;

import android.support.annotation.NonNull;

import java.util.List;

import timber.log.Timber;

/**
 * Created by dug on 5/10/17.
 */

public class DataProjectAddressCombo implements Comparable<DataProjectAddressCombo> {
    public long id;
    public long projectNameId;
    public long addressId;
    String      mProjectName;
    DataAddress mAddress;

    public DataProjectAddressCombo(long projectNameId, long addressId) {
        this.projectNameId = projectNameId;
        this.addressId = addressId;
    }

    public DataProjectAddressCombo(long rowId, long projectNameId, long addressId) {
        this.id = rowId;
        this.projectNameId = projectNameId;
        this.addressId = addressId;
    }

    public void reset(long projectNameId, long addressId) {
        this.projectNameId = projectNameId;
        this.addressId = addressId;
        mProjectName = null;
        mAddress = null;
    }

    public String getProjectName() {
        if (mProjectName == null) {
            mProjectName = TableProjects.getInstance().queryProjectName(projectNameId);
            if (mProjectName == null) {
                Timber.e("Could not find project ID=" + projectNameId);
            }
        }
        return mProjectName;
    }

    public DataProject getProject() {
        return TableProjects.getInstance().queryById(projectNameId);
    }

    public DataAddress getAddress() {
        if (mAddress == null) {
            mAddress = TableAddress.getInstance().query(addressId);
            if (mAddress == null) {
                Timber.e("Could not find address ID=" + addressId);
                Timber.e("ADDRESSES=" + TableAddress.getInstance().toString());
            }
        }
        return mAddress;
    }

    public String getAddressLine() {
        DataAddress address = getAddress();
        if (address != null) {
            return address.getLine();
        }
        return "Invalid";
    }

    public String getCompanyName() {
        DataAddress address = getAddress();
        if (address != null) {
            return address.company;
        }
        return null;
    }

    public String getHintLine() {
        StringBuffer sbuf = new StringBuffer();
        sbuf.append(getProjectName());
        sbuf.append("\n");
        sbuf.append(getCompanyName());
        return sbuf.toString();
    }

    public List<DataEntry> getEntries() {
        return TableEntry.getInstance().queryForProjectAddressCombo(id);
    }

    @Override
    public int compareTo(@NonNull DataProjectAddressCombo o) {
        String name = getProjectName();
        String otherName = o.getProjectName();
        if (name != null && otherName != null) {
            return name.compareTo(otherName);
        }
        if (name == null && otherName == null) {
            return 0;
        }
        return 1;
    }

    public boolean hasValidState() {
        return getAddress().hasValidState();
    }

    public void fix() {
        getAddress().fix();
    }

    @Override
    public String toString() {
        StringBuilder sbuf = new StringBuilder();
        sbuf.append("ID=");
        sbuf.append(id);
        sbuf.append(", PROJID=");
        sbuf.append(projectNameId);
        sbuf.append(" [");
        sbuf.append(getProjectName());
        sbuf.append("] ADDRESS=");
        sbuf.append(addressId);
        if (getAddress() != null) {
            sbuf.append(" [");
            sbuf.append(getAddress().getBlock());
            sbuf.append("]");
        }
        return sbuf.toString();
    }
}

File: appMain/src/main/java/com/cartlc/tracker/model/data/DataProjectAddressCombo.kt
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.model.data

import com.cartlc.tracker.model.table.DatabaseTable
import timber.log.Timber

/**
 * Created by dug on 5/10/17.
 */
class DataProjectAddressCombo : Comparable<DataProjectAddressCombo> {

    constructor(db: DatabaseTable, projectNameId: Long, addressId: Long) {
        this.db = db
        this.projectNameId = projectNameId
        this.addressId = addressId
    }

    constructor(db: DatabaseTable, rowId: Long, projectNameId: Long, addressId: Long) {
        this.db = db
        this.id = rowId
        this.projectNameId = projectNameId
        this.addressId = addressId
    }

    var id: Long = 0
    var projectNameId: Long
    var addressId: Long

    private var mAddress: DataAddress? = null
    private var db: DatabaseTable

    val projectDashName: String
        get() {
            val name = db.tableProjects.queryProjectName(projectNameId) ?: return "-"
            val rootProject = name.first
            val subProject = name.second
            if (subProject.isEmpty()) {
                return rootProject
            }
            return "$rootProject - $subProject"
        }

    val isRootProject: Boolean
        get() {
            val name = db.tableProjects.queryProjectName(projectNameId) ?: return false
            return name.second.isEmpty()
        }

    val rootName: String
        get() {
            val name = db.tableProjects.queryProjectName(projectNameId) ?: return ""
            return name.first
        }

    val projectName: Pair<String, String>?
        get() = db.tableProjects.queryProjectName(projectNameId)

    val project: DataProject?
        get() = db.tableProjects.queryById(projectNameId)

    val address: DataAddress?
        get() {
            if (mAddress == null) {
                mAddress = db.tableAddress.query(addressId)
                if (mAddress == null) {
                    Timber.e("Could not find address ID=$addressId")
                }
            }
            return mAddress
        }

    val addressLine: String
        get() {
            val address = address
            return address?.line ?: "Invalid"
        }

    val companyName: String?
        get() {
            val address = address
            return address?.company
        }

    val hintLine: String
        get() {
            val sbuf = StringBuffer()
            sbuf.append(projectName)
            sbuf.append("\n")
            sbuf.append(companyName)
            return sbuf.toString()
        }

    val entries: List<DataEntry>
        get() = db.tableEntry.queryForProjectAddressCombo(id)

    fun reset(projectNameId: Long, addressId: Long) {
        this.projectNameId = projectNameId
        this.addressId = addressId
//        mProjectName = null
        mAddress = null
    }

    override fun compareTo(other: DataProjectAddressCombo): Int {
        val name = projectDashName
        val otherName = other.projectDashName
        return name.compareTo(otherName)
    }

    override fun toString(): String {
        val sbuf = StringBuilder()
        sbuf.append("ID=")
        sbuf.append(id)
        sbuf.append(", PROJID=")
        sbuf.append(projectNameId)
        sbuf.append(" [")
        sbuf.append(projectName)
        sbuf.append("] ADDRESS=")
        sbuf.append(addressId)
        if (address != null) {
            sbuf.append(" [")
            sbuf.append(address!!.block)
            sbuf.append("]")
        }
        return sbuf.toString()
    }


//    val hasValidState: Boolean
//        get() = address?.hasValidState() ?: false
//
//    fun fix(): DataAddress? {
//        address?.let {
//            if (it.fix()) {
//                return it
//            }
//        }
//        return null
//    }
}
